### ISSUE TITLE
In ipython 8.18 (but not ipython 8.16), the current event loop is overwritten in each cell

### ISSUE THREAD

**Issue by danking** [ipython/ipython#14297]
I expected Out[5] to be `True`.

Admittedly, normal users would not do this, but we maintain a library (https://github.com/hail-is/hail) which provides some synchronous primitives that wrap asynchronous primitives. Those synchronous primitives can use an already set event loop, if one exists, but if no event loop is set, they create a fresh loop and use it. They expect that loop to persist across cells because `aiohttp.ClientSession` stashes a reference to the loop and expects that loop to be the running loop for each subsequent method call.

What can I do, as a library maintainer, to tell ipython to either (1) provide me an event loop or (2) not overwrite an event loop I set?

```ipython3
# ipython
Python 3.10.9 (main, Jan 11 2023, 09:18:18) [Clang 14.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.20.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import asyncio

In [2]: x = asyncio.get_event_loop()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[2], line 1
----> 1 x = asyncio.get_event_loop()

File ~/miniconda3/lib/python3.10/asyncio/events.py:671, in BaseDefaultEventLoopPolicy.get_event_loop(self)
    668     self.set_event_loop(self.new_event_loop())
    670 if self._local._loop is None:
--> 671     raise RuntimeError('There is no current event loop in thread %r.'
    672                        % threading.current_thread())
    674 return self._local._loop

RuntimeError: There is no current event loop in thread <_MainThread(MainThread, started 8706367616)>.

In [3]: y = asyncio.new_event_loop()
   ...: asyncio.set_event_loop(y)

In [4]: await asyncio.sleep(1)
   ...: x = asyncio.get_event_loop()
x == y
In [5]: x == y
Out[5]: False

```

**Comment by JasperSui** [comment]
Hey @danking, I'm the author of [fastapi-injectable](https://github.com/JasperSui/fastapi-injectable), and I'm facing the same issue as you while using IPython with my library.

After looking into IPython's internals, I found that we can retrieve the event loop used when executing `await {COROUTINE}` in the IPython terminal. This allows us to get the loop beforehand and instantiate objects that assume a consistent event loop. For example, `aiohttp.ClientSession` requires both instantiation and subsequent function calls to be on the same event loop.

I've tested this approach on IPython 8.16 and 8.30, and it works without issues:

```python3
from IPython.core.async_helpers import get_asyncio_loop
ipython_loop = get_asyncio_loop()
# Now, you can do whatever you want
# Either: 
# asyncio.set_event_loop(ipython_loop)
# Or pass the ipython_loop to your library and instantiate those objects
```

In short, as long as we ensure that the event loop used during instantiation is the same as the one IPython uses to run `await {COROUTINE}`, everything should work fine.

Hope this helps!

**Comment by ul-gh** [comment]
I somehow like to test my async code in IPython and yes, it would be nice if the event loop could be debugged in IPython or just not be overwritten every time when a cell stops executing.
